tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java

```diff
@@ -224,33 +226,62 @@ private GraphOperation graphOp(Operand<?> operand) {
+   * solely on inputs)
+   * @see #completeSubgraph(Set, Set, Set, Set)
+   */
+  public Set<GraphOperation> completeSubgraph(Set<Operand<?>> inputs, Set<Operand<?>> outputs,
+      boolean allowNoInputBodyOps) {
+    return completeSubgraph(inputs, outputs, null, allowNoInputBodyOps ? Collections.emptySet() : null);
+  }
+
+  /**
+   * Finds the operations used to produce {@code outputs} from {@code inputs}, or throws if that is not possible.
+   * Includes control dependencies.
+   *
+   * If both {@code allowedNoInputBodyOps} and {@code forbiddenNoInputBodyOps} are {@code null}, forbids 0-input ops in
+   * the body. To allow all ops in the body, use {@code null} for {@code allowedNoInputBodyOps} and the empty set for
+   * {@code forbiddenNoInputBodyOps}.
    *
-   * @param inputs the inputs of the subgraph.  Must be from single output ops.
-   * @param outputs the outputs of the subgraph
-   * @param allowedBodyOps types of ops to allow as 0-input ops in the body.  Allows all (except {@code
-   * forbiddenBodyOps}) if null.
-   * @param forbiddenBodyOps types of ops to never allow as 0-input ops in the body.  Forbids all (except {@code
-   * allowedBodyOps}) if null.
+   * @param inputs the inputs of the subgraph.  Must be from single output ops.  May not be null.
+   * @param outputs the outputs of the subgraph.  May not be null.
+   * @param allowedNoInputBodyOps types of ops to allow as 0-input ops in the body.  Allows all (except {@code
+   * forbiddenNoInputBodyOps}) if null.
+   * @param forbiddenNoInputBodyOps types of ops to never allow as 0-input ops in the body.  Forbids all (except {@code
+   * allowedNoInputBodyOps}) if null.
    * @return the set of operations needed to calculate outputs from inputs, including outputs and inputs
    * @throws IllegalStateException if outputs depends on ops outside of the subgraph (i.e. is not calculable based
    * solely on inputs)
+   * @see #completeSubgraph(Set, Set, boolean)
    */
   public synchronized Set<GraphOperation> completeSubgraph(Set<Operand<?>> inputs, Set<Operand<?>> outputs,
-      Set<String> allowedBodyOps, Set<String> forbiddenBodyOps) {
+      Set<String> allowedNoInputBodyOps, Set<String> forbiddenNoInputBodyOps) {
 
-    if (forbiddenBodyOps != null && allowedBodyOps != null) {
-      for (String t : forbiddenBodyOps) {
-        if (allowedBodyOps.contains(t)) {
+    if (forbiddenNoInputBodyOps != null && allowedNoInputBodyOps != null) {
+      for (String t : forbiddenNoInputBodyOps) {
+        if (allowedNoInputBodyOps.contains(t)) {
           throw new IllegalArgumentException("Can't allow and forbid op type " + t + ".");
         }
       }
     }
 
+    if (inputs == null) {
+      throw new IllegalArgumentException("Inputs can't be null.");
+    }
+
+    if (outputs == null) {
+      throw new IllegalArgumentException("Outputs can't be null.");
+    }
+
     Queue<GraphOperation> currents = new LinkedList<>();
```

Comment:
Given this is an `ArrayDeque` in the other new methods, maybe this one should be an `ArrayDeque` not a `LinkedList`

Reply:
They should be LinkedList too, I think, for performance.  All I'm doing is front removal and back append, no indexing.

Reply:
The performance of `LinkedList` is bad due to having to do two pointer chases to find any given object and thus likely induce two cache misses (one for the node and one for the object itself). I'd expect `ArrayDeque` to be faster given it is based on a circular array, so it's likely to have the array always in cache when it's being traversed.

However performance isn't likely to be an issue in this code, so provided they are all the same I don't care.

Reply:
Polling takes about the same amount of dereferences, either `LinkedList -> Node -> Object` or `ArrayDeque -> backing array -> Object`.